akka-docs/src/main/paradox/stream/stream-context.md

```diff
@@ -0,0 +1,71 @@
+# Context Propagation
+
+It can be convenient to attach metadata to each element in the stream.
+
+For example, when reading from an external data source it can be
+useful to keep track of the read offset, so it can be marked as processed
+when the element reaches the @apidoc[Sink].
+
+For this use case we provide the @apidoc[SourceWithContext] and
+@apidoc[FlowWithContext] variations on @apidoc[Source] and
+@apidoc[Flow].
+
+Essentially, a @apidoc[FlowWithContext] is just a @apidoc[Flow] that
+contains @scala[tuples]@java[pairs] of element and context, but the
+advantage is in the operators: most operators on @apidoc[FlowWithContext]
+will work on the element rather than on the @scala[tuple]@java[pair],
+allowing you to focus on your application logic rather without worrying
+about the context.
+
+## Restrictions
+
+Not all operations that are available on @apidoc[Flow] are also available
+on @apidoc[FlowWithContext]. This is intentional: in the use case of
+keeping track of a read offset, if the @apidoc[FlowWithContext] was
+allowed to arbitrarily filter and reorder the stream, the @apidoc[Sink]
+would have no way to determine whether an element was skipped
+or merely reordered and still in flight.
+
+For this reason, @apidoc[FlowWithContext] allows filtering operations
+(such as `filter`, `filterNot`, `collect`, etc) and grouping operations
+(such as `grouped`, `sliding`, etc) but not reordering operations
+(such as `mapAsyncUnordered` and `statefulMapConcat`). Finally,
+also 'one-to-n' operations such as `mapConcat` are allowed.
+
+Filtering operations will drop the context along with dropped elements,
+while grouping operations will keep all contexts from the elements in
+the group. Streaming one-to-many operations such as `mapConcat`
+associate the original context with each of the produced elements.
```

Comment:
Should we add an extra note here saying that one-to-many can be dangerous when side-effecting? 
In case of saving an offset, this is the easiest way to screw up as some elements may not be processed while the offset is saved.

Reply:
Not sure - while I agree that is important to understand, I think discussing it here would make the docs harder to follow...